### ISSUE TITLE
Received signal Segmentation fault while using multithread to run alter add column

### ISSUE THREAD

**Issue by laoguanjie** [yandex/ClickHouse#2480]
i use multithread to add columns to a table like
```
alter table local_1.metriczk add column `system.filesystem.files` String
```

but clickhouse crash with info like 
```
2018.06.05 17:45:45.704610 [ 4415 ] <Warning> local_1.metriczk (StorageReplicatedMergeTree): /clickhouse/tables/1/metriczk/2018-06-05/02/hits/columns changed before this ALTER finished; overlapping ALTER-s are fine but use caution with nontransitive changes
2018.06.05 17:45:45.827694 [ 146 ] <Warning> local_1.metriczk (StorageReplicatedMergeTree): /clickhouse/tables/1/metriczk/2018-06-05/02/hits/columns changed before this ALTER finished; overlapping ALTER-s are fine but use caution with nontransitive changes
2018.06.05 17:45:45.935667 [ 470 ] <Error> BaseDaemon: ########################################
2018.06.05 17:45:45.935705 [ 470 ] <Error> BaseDaemon: (from thread 4415) Received signal Segmentation fault (11).
2018.06.05 17:45:45.935713 [ 470 ] <Error> BaseDaemon: Address: 0xfffffffffffffff8
2018.06.05 17:45:46.103805 [ 470 ] <Error> BaseDaemon: 0. clickhouse-server(auto ext::collection_cast<DB::NamesAndTypesList, boost::range::joined_range<DB::NamesAndTypesList const, DB::NamesAndTypesList const> >(boost::range::joined_range<DB::NamesAndTypesList const, DB::NamesAndTypesList const> const&)+0x61) [0x416e531]
2018.06.05 17:45:46.103833 [ 470 ] <Error> BaseDaemon: 1. clickhouse-server(DB::ITableDeclaration::getColumnsList() const+0x5f) [0x4165bef]
2018.06.05 17:45:46.103857 [ 470 ] <Error> BaseDaemon: 2. clickhouse-server(DB::AlterCommands::validate(DB::IStorage*, DB::Context const&)+0x35) [0x44a10b5]
2018.06.05 17:45:46.103867 [ 470 ] <Error> BaseDaemon: 3. clickhouse-server(DB::InterpreterAlterQuery::execute()+0x145) [0x4440535]
2018.06.05 17:45:46.103875 [ 470 ] <Error> BaseDaemon: 4. clickhouse-server() [0x40f3bdf]
2018.06.05 17:45:46.103888 [ 470 ] <Error> BaseDaemon: 5. clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, DB::Context&, std::function<void (std::string const&)>)+0x112) [0x40f4fb2]
2018.06.05 17:45:46.103901 [ 470 ] <Error> BaseDaemon: 6. clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x2b63) [0x2609763]
2018.06.05 17:45:46.103912 [ 470 ] <Error> BaseDaemon: 7. clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x377) [0x260bae7]
2018.06.05 17:45:46.103934 [ 470 ] <Error> BaseDaemon: 8. clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x2a2) [0x47f2632]
2018.06.05 17:45:46.103944 [ 470 ] <Error> BaseDaemon: 9. clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x47ee56f]
2018.06.05 17:45:46.103954 [ 470 ] <Error> BaseDaemon: 10. clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x110) [0x47eebd0]
2018.06.05 17:45:46.103965 [ 470 ] <Error> BaseDaemon: 11. clickhouse-server(Poco::PooledThread::run()+0x77) [0x4a08f57]
2018.06.05 17:45:46.103982 [ 470 ] <Error> BaseDaemon: 12. clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x38) [0x4a066b8]
2018.06.05 17:45:46.103995 [ 470 ] <Error> BaseDaemon: 13. clickhouse-server() [0x4b027cf]
2018.06.05 17:45:46.104010 [ 470 ] <Error> BaseDaemon: 14. /lib64/libpthread.so.0(+0x7e25) [0x7f76b35dde25]
```
local_1.metriczk 's table info:

```sql
CREATE TABLE local_1.metriczk (
  `kafka.partition.topic.name` String,
   tenant_id Float64,
   uuid String,
  `metricset.host` String,
  `kafka.partition.partition.leader` Float64,
   logtype String,
   keeptype String,
   hostname String,
  `kafka.partition.offset.newest` Float64,
   appname String,
   host String,
   tag String,
  `kafka.partition.partition.id` Float64,
   timestamp String,
  `kafka.partition.partition.replica` Float64,
  `kafka.partition.partition.insync_replica` String,
   at_date Date,
  `kafka.partition.broker.id` Float64,
  `kafka.partition.broker.address` String,
  `metricset.module` String,
  `metricset.rtt` Float64,
   message String,
  `metricset.name` String,
   agent_type String,
   at_timestamp Float64,
  `kafka.partition.offset.oldest` Float64,
  `system.load.1` Float64,
  `system.load.5` Float64,
  `system.load.cores` Float64,
  `system.load.15` Float64,
  `system.load.norm.15` Float64,
  `system.load.norm.5` Float64,
  `system.load.norm.1` Float64,
  `zookeeper.mntr.latency.min` Float64,
  `zookeeper.mntr.latency.avg` Float64,
  `zookeeper.mntr.watch_count` Float64,
  `zookeeper.mntr.server_state` String,
  `zookeeper.mntr.outstanding_requests` Float64,
  `zookeeper.mntr.open_file_descriptor_count` Float64,
  `zookeeper.mntr.ephemerals_count` Float64,
  `zookeeper.mntr.num_alive_connections` Float64,
  `zookeeper.mntr.max_file_descriptor_count` Float64,
  `zookeeper.mntr.packets.received` Float64,
  `zookeeper.mntr.version` String,
  `zookeeper.mntr.znode_count` Float64,
  `zookeeper.mntr.packets.sent` Float64,
  `zookeeper.mntr.approximate_data_size` Float64,
  `zookeeper.mntr.latency.max` Float64,
  `system.uptime.duration.ms` Float64,
  `system.core.user.pct` Float64,
  `system.core.irq.pct` Float64,
  `system.core.nice.pct` Float64,
  `system.network.out.errors` Float64,
  `system.core.idle.pct` Float64,
  `system.network.in.bytes` Float64,
  `system.core.id` Float64,
  `system.network.in.dropped` Float64,
  `system.network.out.dropped` Float64,
  `system.core.softirq.pct` Float64,
  `system.core.system.pct` Float64,
  `system.network.out.packets` Float64,
  `system.core.iowait.pct` Float64,
  `system.network.out.bytes` Float64,
  `system.network.in.packets` Float64,
  `system.core.steal.pct` Float64,
  `system.cpu.system.pct` Float64,
  `system.network.in.errors` Float64,
  `system.cpu.total.pct` Float64,
  `system.network.name` String,
  `system.cpu.irq.pct` Float64,
  `system.fsstat.total_size.free` Float64,
  `system.cpu.cores` Float64,
  `system.fsstat.total_files` Float64,
  `system.cpu.iowait.pct` Float64,
  `system.fsstat.total_size.total` Float64,
  `system.cpu.softirq.pct` Float64,
  `system.fsstat.count` Float64,
  `system.cpu.idle.pct` Float64,
  `system.memory.used.pct` Float64,
  `system.cpu.nice.pct` Float64,
  `system.memory.swap.total` Float64,
  `system.cpu.steal.pct` Float64,
  `system.memory.actual.free` Float64,
  `system.cpu.user.pct` Float64,
  `system.memory.total` Float64,
  `system.diskio.iostat.read.request.merges_per_sec` Float64,
  `system.memory.swap.used.pct` Float64,
  `system.diskio.write.bytes` Float64,
  `system.memory.swap.used.bytes` Float64,
  `system.diskio.iostat.write.request.merges_per_sec` Float64,
  `system.memory.free` Float64,
  `system.diskio.io.time` Float64,
  `system.diskio.read.count` Float64,
  `system.memory.swap.free` Float64,
  `system.diskio.iostat.service_time` Float64,
  `system.memory.actual.used.pct` Float64,
  `system.diskio.iostat.write.request.per_sec` Float64,
  `system.memory.actual.used.bytes` Float64,
  `system.memory.used.bytes` Float64,
  `system.diskio.read.bytes` Float64,
  `system.diskio.write.time` Float64,
  `system.diskio.iostat.await` Float64,
  `system.diskio.iostat.write.per_sec.bytes` Float64,
  `system.diskio.iostat.read.request.per_sec` Float64,
  `system.filesystem.mount_point` String,
  `system.diskio.iostat.read.per_sec.bytes` Float64,
  `system.filesystem.files` Float64,
  `system.diskio.write.count` Float64,
  `system.diskio.iostat.queue.avg_size` Float64,
  `system.filesystem.device_name` String,
  `system.filesystem.used.pct` Float64,
  `system.filesystem.free_files` Float64,
  `system.filesystem.used.bytes` Float64,
  `system.filesystem.type` String,
  `system.filesystem.free` Float64,
  `system.filesystem.available` Float64,
  `system.filesystem.total` Float64,
  `system.diskio.read.time` Float64,
  `system.diskio.name` String,
  `system.diskio.iostat.busy` Float64,
  `system.diskio.iostat.request.avg_size` Float64,
  `system.fsstat.total_size.used` Float64,
  `error.message` String
 ) ENGINE = ReplicatedMergeTree('/clickhouse/tables/1/metriczk/2018-06-05/{shard}/hits', '{replica}')
 PARTITION BY concat(toString(at_date), keeptype)
 ORDER BY (at_date, at_timestamp, intHash64(toInt64(at_timestamp)))
 SAMPLE BY intHash64(toInt64(at_timestamp))
 SETTINGS index_granularity = 8192
```

the table has 123 columns.

is it a bug?

**Comment by KochetovNicolai** [comment]
Segmentation fault is a bug, but I can't reproduce on the last version.
I've created table with the same structure on cluster. When I run alter query a get:
```
Code: 44. DB::Exception: Received from localhost:9004, ::1. DB::Exception: Cannot add column system.filesystem.files: column with this name already exists.
```
After I'd added `system.filesystem.files2` column no error has happened.

**Comment by laoguanjie** [comment]
@KochetovNicolai thx your reply:)

i use version 1.1.54381, and it's easy to reproduce.

 i test ch with two points:
> 1、100 threads add columns concurrently.
2、column which these threads try to add may by the same, or not.

i use clickhouse-jdbc to try it, and my code like


```
ClickhouseClient client = new ClickhouseClient("localhost:8123");
//create 100 threads
MultiThread threads = new MultiThread.create(100);
int count = 10000;
for(int i = 0; i < count; i++) {
    threads.exec(new job() {
        	@Override
            public void process() {

                //use System.currentTimeMillis()%2000 to create columns which may have conflict.
                String column = "system.process.cgroup.cpuacct.percpu." + System.currentTimeMillis()%2000;

                String template = "alter table local_1.metriczk add column `{}` String";		

                //create sql like 'alter table local_1.metriczk add column `system.process.cgroup.cpuacct.percpu.312` String'
                String addColumnSql = formatSql(template, column);

                client.exec(sql);
            }
    });
}
```

run it,and then ch crash...

**Comment by filimonov** [comment]
Does it still reproduce on recent versions?

**Comment by alexey-milovidov** [comment]
I'm definitely sure it is fixed at least in version 19.14-testing.
